Added capability of retrieving price info from Taptools

ajuggler opened the pull request:
Most relevant new function is `getTaptoolsPrice` defined in 'Prices.hs' .

ToDo:  Add code to stop bot if price discrepancy is beyond some threshold.

ajuggler commented:
Since Taptools aggregates the price provided by MinSwap (and provided by Maestro), it seems desirable to have the capability of fine-tunning the weights of each Prices Provider.  This commit enables that.

ajuggler commented:
With last commit, `buildPP` now checks that
- Length of list of weights coincides with number of prices providers;
- "Threshold2" is not smaller than "Threshold1".

ajuggler commented:
Successfully tested *Taptools price feed*, as well as these scenarios:
- Maestro price is unavailable (warns but MMBot continues)
- Taptools price is unavailable (warns but MMBot continues)
- Both Maestro and Taptools are unavailable (warns and closes all orders)
- Price mismatch lower than given threshold (continues normally; since price average changed, MMBot updates orders accordingly)
- Price mismatch larger than given threshold (closes all orders)

*Note:*  as mechanism to flag changes while MMBot is running, the following files can be updated:
- offset_price.txt (initial content: 0)
- maestro_available.txt (initial content: True)
- taptools_available.txt (initial content: True)

ajuggler commented:
"Flag files" are:
- 'offset_price.txt':  content of this file is artificially added to Maestro's price.
- 'maestro_available.txt':  if its content is `False`, Maestro is artificially unavailable.
- 'taptools_available.txt':  it its content is `False`, Taptools is artificially unavailable.

If *flag files* are not present, default values (zero, True and True, respectively) are assumed.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Utils.hs:
```diff
@@ -53,3 +66,132 @@ instance HasPartialOrderNftScript DEXInfo where
+
+    withHeaders :: Request -> IO Request
+    withHeaders req = do
+      return $ req { requestHeaders = ("x-api-key", encodeUtf8 apiKey') :
```
https://docs.servant.dev/en/stable/tutorial/ApiType.html#request-headers

sourabhxyz replied:
@ajuggler I notice that it is still not handled for?

ajuggler replied:
This is done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -65,6 +129,279 @@ data MMToken = MMToken
+  deriving (FromJSON, ToJSON) via CustomJSON '[FieldLabelModifier '[CamelToSnake]] MaestroPairOverride
+
+data TaptoolsPairOverride = TaptoolsPairOverride
+  { ttpoAsset            ∷ !String,
```
`GYAssetClass`

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Utils.hs:
```diff
@@ -53,3 +66,132 @@ instance HasPartialOrderNftScript DEXInfo where
+                      | CRes1d | CRes1w  | CRes1mo
+                      deriving stock (Eq, Ord, Show, Generic)
+
+instance FromJSON CommonResolution
```
Needs modification to match convention (snake case).

sourabhxyz replied:
And to remove `CRes` in json value, look at `StripPrefix`.

sourabhxyz replied:
I am also fine with getting rid of common resolution altogether as having it implies an assumption that all price providers must have resolution.

ajuggler replied:
Originally I thought that having a common resolution was crucial.  But testing shows that last closing price appears to be independent of resolution.

ajuggler replied:
Update:  as discussed, further experimentation seems to show that closing price does depend on the resolution, at least for Taptools.  For example, there is mismatch between 1 hour and 1 day resolution for "iBTC".  Hence I am keeping "common resolution" for now.

ajuggler replied:
> And to remove `CRes` in json value, look at `StripPrefix`.

My current solution is:
```haskell
instance FromJSON CommonResolution where
    parseJSON (String v) =
      let kvm = Map.fromList [ ("5m", CRes5m), ("15m", CRes15m), ("30m", CRes30m), ("1h", CRes1h),
                               ("4h", CRes4h), ("1d", CRes1d), ("1w", CRes1w), ("1mo", CRes1mo) ]
      in case Map.lookup v kvm of
        Nothing   -> typeMismatch "CommonResolution" ""
        Just cres -> pure cres
    parseJSON invalid    = typeMismatch "CommonResolution" invalid
```

I tried 
```haskell
  deriving (FromJSON) via CustomJSON '[FieldLabelModifier '[StripPrefix Abc]] CommonResolution
```
but `StripPrefix` seems to only work in the context of a record datatype.

ajuggler replied:
> Needs modification to match convention (snake case).

Since we are not deriving `ToJSON`, I don't think we need to worry about snake case here.

ajuggler replied:
Got rid of "common resolution"!

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Utils.hs:
```diff
@@ -53,3 +66,132 @@ instance HasPartialOrderNftScript DEXInfo where
+    numerator   = sum $ map (\(x, y) -> x * y) wps
+    denominator = sum $ map fst wps
+
+relStdDev :: [Double] -> Double
```
`NonEmpty`.

ajuggler replied:
This is done.

sourabhxyz commented on .gitignore:
```diff
@@ -25,3 +25,6 @@ cabal.project.local~
 *.skey
 var/
 .envrc
+my-atlas-config-maestro.json
```
In local environment you can put these under `var` folder or perhaps `secrets` folder with `secrets` added to `.gitignore`.

sourabhxyz replied:
I would suggest omitting `my-preprod-maker-bot-config-gens-v2.yaml` etc from `.gitignore`.

ajuggler replied:
Done.  Moved all "secret" files to folder "secrets".

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/MakerBot.hs:
```diff
@@ -26,56 +27,70 @@ import           System.Exit
   }
 
+data MBFret = MBReady | MBSpooked1
+                        { mbsRelax1 ∷ !Int
```
`mbs1` prefix.

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Utils.hs:
```diff
@@ -53,3 +66,132 @@ instance HasPartialOrderNftScript DEXInfo where
+-------------------------------------------------------------------------------
+
+-- | Taptools time resolution for OHLC Candles
+data TtResolution = TtRes3m | TtRes5m | TtRes15m | TtRes30m | TtRes1h | TtRes2h | TtRes4h | TtRes12h
```
Likewise, remove `TtRes` prefix for json instance.

ajuggler replied:
Done for `FromJSON` instance.  Will also implement `ToJSON` instance.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Utils.hs:
```diff
@@ -53,3 +66,132 @@ instance HasPartialOrderNftScript DEXInfo where
+  show = let kvm = Map.fromList [ (TtRes3m, "3m"), (TtRes5m, "5m"), (TtRes15m, "15m"), (TtRes30m, "30m")
+                                , (TtRes1h, "1h"), (TtRes2h, "2h"), (TtRes4h, "4h"), (TtRes12h, "12h")
+                                , (TtRes1d, "1d"), (TtRes3d, "3d"), (TtRes1w, "1w"), (TtRes1mo, "1M") ]
+         in  fromJust . flip Map.lookup kvm
```
Use of `fromJust` appears unnecessary if done via pattern match. Have a look on how I wrote show instance for resolution [here](https://github.com/maestro-org/haskell-sdk/blob/609515b3f8861efaa17adf4e599aa1b845775a6d/src/Maestro/Types/V1/DefiMarkets.hs#L50-L53).

ajuggler replied:
However I don't think `fromJust` is problematic here, meaning that it does not lead to a partial function.  Because every possible value of `TtResolution` is a key in the given key-value-map `kvm`.

sourabhxyz replied:
If a new data constructor is added, it would create issue.

ajuggler replied:
You're right.  This now has been fixed using pattern match.  I could not use the neat trick [`toConstr >>> show >>> drop 5`](https://github.com/maestro-org/haskell-sdk/blob/609515b3f8861efaa17adf4e599aa1b845775a6d/src/Maestro/Types/V1/DefiMarkets.hs#L50-L53) because pattern breaks in the month case (unless I change the constructor name, but I don't want to do that for consistency with Maestro constructor's name).

ajuggler replied:
Done using pattern match.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Utils.hs:
```diff
@@ -53,3 +66,132 @@ instance HasPartialOrderNftScript DEXInfo where
+
+type TtAPI =
+  "token" :> "ohlcv" :> QueryParam "unit" TtUnit
+                     :> QueryParam "interval" TtInterval
```
isn't "interval" the `TtResolution`? Why use `String` here?

ajuggler replied:
What I get from your comment is that `QueryParam` is able to use the `ToJSON` instance to parse `TtResolution` as a string, and that therefore I should define the corresponding instance.  Then I could use directly `TtResolution` as an argument of `QueryParam`, instead of `TtInterval`.  Am I understanding you correctly?

sourabhxyz replied:
> What I get from your comment is that QueryParam is able to use the ToJSON instance...

No, please have a look at [this section on `FromHttpApiData`/`ToHttpApiData`](https://docs.servant.dev/en/stable/tutorial/Server.html#the-fromhttpapidata-tohttpapidata-classes).

ajuggler replied:
Done:  declared `instance ToHttpApiData TtResolution where` using pattern match and removed `TtInterval`.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Utils.hs:
```diff
@@ -53,3 +66,132 @@ instance HasPartialOrderNftScript DEXInfo where
+                   }
+
+taptoolsBaseUrl :: BaseUrl
+taptoolsBaseUrl = BaseUrl Http "openapi.taptools.io" 80 "api/v1"
```
Use https instead of http.

ajuggler replied:
Done:  changed `Http` to `Https` and 80 to 443.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Utils.hs:
```diff
@@ -53,3 +66,132 @@ instance HasPartialOrderNftScript DEXInfo where
+taptoolsBaseUrl = BaseUrl Http "openapi.taptools.io" 80 "api/v1"
+
+priceFromTaptools :: Maybe TtUnit -> Maybe TtInterval -> Maybe Int -> ClientEnv -> IO (Either ClientError [TtOHLCV])
+priceFromTaptools mbUnit mbInterval mbNumIntervals = runClientM (getTtOHLCV mbUnit mbInterval mbNumIntervals)
```
To me it appears that all of Taptools client code thingy should be part of either `Prices.hs` or a dedicated module, `GeniusYield.MarketMaker.Prices.TapTools`.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Utils.hs:
```diff
@@ -53,3 +66,132 @@ instance HasPartialOrderNftScript DEXInfo where
+priceFromTaptools mbUnit mbInterval mbNumIntervals = runClientM (getTtOHLCV mbUnit mbInterval mbNumIntervals)
+
+-- | Get @GYAssetClass@ from string 'policyid.tokenname' in Hex format.
+taptoolsParseAsset :: String -> Maybe GYAssetClass
```
To be omitted and config to rather have `GYAssetClass`.

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -52,11 +57,70 @@ import           Maestro.Types.V1
+data PriceCommonCfg = PriceCommonCfg
+  { pccCommonResolution       ∷ !CommonResolution,
+    pccNetworkId              ∷ !GYNetworkId,
+    pccPricesProvidersWeights ∷ ![Int],   -- ^ Corresponding to each @PricesProviderCfg@
```
Since weight ties to each provider, I would rather prefer `pcPricesProvderCfgs` to be a list of tuples instead where second element would be weight.

ajuggler replied:
Or rather, weight should be one of the required fields for each prices provider.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -52,11 +57,70 @@ import           Maestro.Types.V1
+
+data PricesProviderCfg =
+    MaestroConfig
+    { mcApiKey             ∷ !(Confidential Text),
```
This way of defining record leads to partial fields, i.e., `ttcApiKey` would throw an error if data constructor was `MaestroConfig`. Suggestion is to have dedicated types for `MaestroConfig` and `TaptoolsConfig`.

sourabhxyz replied:
Yes (as discussed in our meeting earlier).

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices/Taptools.hs:
```diff
@@ -0,0 +1,91 @@
+                  | TtRes1d | TtRes3d | TtRes1w  | TtRes1mo
+                  deriving stock (Eq, Ord, Generic)
+
+instance FromJSON TtResolution where
```
Use automatic `FromJSON` derivation, like done in other places: https://github.com/maestro-org/haskell-sdk/blob/609515b3f8861efaa17adf4e599aa1b845775a6d/src/Maestro/Types/V1/Common.hs#L89-L91.

ajuggler replied:
Done.

sourabhxyz replied:
Have a look at automatic derivation here: https://github.com/maestro-org/haskell-sdk/blob/609515b3f8861efaa17adf4e599aa1b845775a6d/src/Maestro/Types/V1/DefiMarkets.hs#L45-L48.

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Strategies.hs:
```diff
@@ -185,6 +188,25 @@ filterOwnOrders mmts users allOrders =
+  pe <- priceEstimate pp mmTokenPair
+  case pe of
+    PriceSourceFail p -> do
+      logWarn providers $ "One prices provider failed"  -- TODO: give details of failing provider
```
Do you plan to tackle this todo in this PR?

sourabhxyz replied:
A warning log has been made for this case but not for others which seem inconsistent.

ajuggler replied:
Done.  Tackled this issue in [this commit](https://github.com/geniusyield/market-maker/pull/73/commits/6e78e20a3ffbb434a399bd58c7f5307a5e8d31d7).

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -52,11 +58,70 @@ import           Maestro.Types.V1
+      mcResolutionOverride ∷ !(Maybe Maestro.Resolution),
+      mcDex                ∷ !Dex,
+      mcPairOverride       ∷ !(Maybe MaestroPairOverride),
+      mcOffsetFilePath     ∷ !(Maybe FilePath),  -- TESTING
```
Now that you are using MVars for your testing, do you still need this filepath based logic?

ajuggler replied:
Kept files as mechanism for passing state only for "Mock prices provider".  Removed filepaths in config for Maestro and Taptools.

ajuggler replied:
Now state of mock prices providers is handled with MVars.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -65,6 +130,277 @@ data MMToken = MMToken
+  deriving stock (Show, Generic)
+  deriving (FromJSON, ToJSON) via CustomJSON '[FieldLabelModifier '[CamelToSnake]] TaptoolsPairOverride
+
+data PricesProvidersAggregator = PPA
```
Isn't this same as `PriceConfigV2`?

ajuggler replied:
Yes.  In my mind I was trying to separate the data types used for configuration and the ones used for constructing the `PricesProviders` with function `buildPP`.  But now that you point it out, there is no advantage in this strict separation.  Will use `PriceConfigV2` in place of `PricesProvidersAggregator`.

sourabhxyz replied:
`PPA` can be handled using pattern synonym.

ajuggler replied:
Managed to successfully work with a pattern synonym with the code:
```haskell
{-# COMPLETE PPA #-}
-- | Prices providers aggregator
pattern PPA :: PriceCommonCfg -> NonEmpty PricesProviderCfg -> PriceConfigV2
pattern PPA {ppaCommonCfg, ppaPricesCluster} = PriceConfigV2 { pcPriceCommonCfg = ppaCommonCfg, pcPricesProviderCfgs = ppaPricesCluster }
```

But in the end I introduced again datatype `PricesProvidersAggregator` to distinguish the data in `PricesProviders` (which is the result of `buildPP`) from the configuration data.  This was necessary to manage the MVar's that store the state of the mock prices providers in `PricesProviders`.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -65,6 +130,277 @@ data MMToken = MMToken
+
+  let pair = TaggedText pairName
+
+  ohlInfo ← handleMaestroError (functionLocationIdent <> " - fetching price from pair") <=< try $ pricesFromDex env mcDex pair (Just res) (Just Descending)
```
`handleMaestroError` would be throwing error and I suppose you wanted to return `SourceUnavailable`?

ajuggler replied:
Indeed.  I have modified accordingly.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/MakerBot.hs:
```diff
@@ -115,19 +130,117 @@ buildAndSubmitActions user@User {uColl, uStakeCred} providers netId ua di = flip
+
+          UACSpooked1   → do
+            put MBSpooked1 { mbs1Relax = 0, mbs1Worse = 0 }
+            lift $ gyLogInfo providers logNS "Closed all orders due to price mismatch among Prices Providers"
```
Closing* as orders aren't yet closed.
Put a log again after `cancelAllOrders'` successfully returns.

ajuggler replied:
Done:  changed "Closed" to "Closing".  Function `cancelAllOrders'` already logs successful return with "No more orders to cancel!"

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/MakerBot.hs:
```diff
@@ -115,19 +130,117 @@ buildAndSubmitActions user@User {uColl, uStakeCred} providers netId ua di = flip
+
+          UACSpooked2 e → do
+            put MBSpooked2 { mbs2Relax = 0 }
+            lift $ gyLogWarning providers logNS $ "Closed all orders due to: " ++ e
```
Closing*

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -52,11 +58,70 @@ import           Maestro.Types.V1
+    pccNetworkId              ∷ !GYNetworkId,
+    pccPriceDiffThreshold1    ∷ !Double,  -- ^ Triggers "mildly spooked"
+    pccPriceDiffThreshold2    ∷ !Double,  -- ^ Triggers "very spooked"; Threshold2 >= Threshold1
+    pccRespiteDelayFactor     ∷ !Double,  -- ^ Thread delay multiplier determining the respite delay when  "spooked"
```
I understand it's purpose but I wonder whether this was a design requirement. I have a feeling that `mbDelay` along with `pccAfterExitRelaxAim1` etc. is sufficient.

ajuggler replied:
Useful for expedite testing, but let us ponder its removal.

sourabhxyz replied:
Let's make it under `Maybe` or omitted for.

ajuggler replied:
Done, it is now under `Maybe`.

ajuggler commented:
I hard-coded configuration files' path in lines 46 & 47 of 'geniusyield-market-maker/test/Main.hs' because I could not figure out how to pass these as arguments using the `cabal test` command.  This of course has to be changed.

ajuggler commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -65,6 +130,277 @@ data MMToken = MMToken
+    PCVersion2 (PriceConfigV2 {..}) → do
+      let ppaCommonCfg     = pcPriceCommonCfg
+          thresHold1       = pccPriceDiffThreshold1 ppaCommonCfg
+          thresHold2       = pccPriceDiffThreshold1 ppaCommonCfg
```
Done.  Thanks for spotting this typo!

ajuggler commented:
Latest changes:
- I removed all code duplication in testing suite.
- Added a third prices provider, so now `PricesProviderCfg` is:
```haskell
data PricesProviderCfg =
    MaestroPPC MaestroConfig
  | TaptoolsPPC TaptoolsConfig
  | MockPPC MockConfig
```
- The tested scenario is encoded in [`testSequence :: [PPStatus]`](https://github.com/geniusyield/market-maker/blob/e173543ab687f6660218aca186960730f79a8f07/geniusyield-market-maker/test/GeniusYield/Test/MarketMaker/Utils.hs#L49)
- The validation logic is encoded in [`testRequirements :: [[LogData -> Bool]]`](https://github.com/geniusyield/market-maker/blob/e173543ab687f6660218aca186960730f79a8f07/geniusyield-market-maker/test/GeniusYield/Test/MarketMaker/Utils.hs#L61)

Executing `cabal test` resulted in:
```
OK (733.48s)

All 1 tests passed (733.51s)
Test suite pproviders-status-sequence: PASS
Test suite logged to:
/Users/antonio/code/haskell/geniusyield/market-maker/dist-newstyle/build/x86_64-osx/ghc-9.2.8/geniusyield-market-maker-0.3.0/t/pproviders-status-sequence/test/geniusyield-market-maker-0.3.0-pproviders-status-sequence.log
1 of 1 test suites (1 of 1 test cases) passed.
```

ajuggler commented:
Latest commit addresses the following issues:
- If Maestro failed to provide price, execution would stop with an exception instead of relying on the other prices provider(s).
- The warning log message stating that a prices provider failed was not indicating which one.
- There was an inconsistency in where in the code the "prices provider fail" log message was being triggered.

With latest commit, prices provider failure consists of:
1. Warning that "Some prices provider(s) failed: " and a list of those that failed.
2. Logging the exceptions thrown by each failing prices provider.

Function [`logPricesProviderFail ∷ GYProviders → PriceIndicator → IO ()`](https://github.com/geniusyield/market-maker/blob/6e78e20a3ffbb434a399bd58c7f5307a5e8d31d7/geniusyield-market-maker/src/GeniusYield/MarketMaker/Strategies.hs#L435), exported by module `Strategies`, implements the above. 

To elaborate, logging of a prices provider failure can occur if some, but not all, prices providers fail:
- While executing the trading strategy, or
- While on standby (in "spooked mode") waiting to confirm that all prices providers are again in agreement.

ajuggler commented:
I believe this PR is ready to be merged into `main`, and that the following outstanding issues can be worked out while MMBot is being tested on Preprod, as a separate PR.

Outstanding issues:
- [ ] Log each client's exception in the unlikely case that all prices providers fail.  (Currently only a warning message announcing all prices providers failed is logged.)
- [ ] Use CURL to query taptools prices (in order to be able to use https).
- [ ] Consider removing parameter `pccAfterExitRelaxAim1`.
- [ ] Add security measure to make sure that secret taptools key is not logged in case of taptools failure.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices/Taptools.hs:
```diff
@@ -0,0 +1,99 @@
+
+type TtAPI =
+  "token" :> "ohlcv" :> QueryParam "unit" TtUnit
+                     :> QueryParam "interval" TtResolution
```
This is a required query parameter. For reference, you can look at my PR which makes it required: https://github.com/geniusyield/dex-contracts-api/pull/60/files.

ajuggler commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -65,6 +130,277 @@ data MMToken = MMToken
+          thresHold2       = pccPriceDiffThreshold1 ppaCommonCfg
+          ppaPricesCluster = NE.toList pcPricesProviderCfgs
+
+      when (length (pccPricesProvidersWeights ppaCommonCfg) /= length ppaPricesCluster) $
```
Have kept it as it is since weights are passed as a list in configuration.

sourabhxyz commented on geniusyield-market-maker/geniusyield-market-maker.cabal:
```diff
@@ -149,9 +154,66 @@ executable geniusyield-market-maker-exe
     , random
     , random-fu
     , vector
+  ghc-options:      -Wall
```
Likely already set by common `import`'s.

ajuggler replied:
Indeed; done.

sourabhxyz commented on geniusyield-market-maker/geniusyield-market-maker.cabal:
```diff
@@ -149,9 +154,66 @@ executable geniusyield-market-maker-exe
     , random-fu
     , vector
+  ghc-options:      -Wall
+  default-language: Haskell2010
```
Likely already set by common `import`'s.

ajuggler replied:
Indeed; done.

sourabhxyz commented on geniusyield-market-maker/geniusyield-market-maker.cabal:
```diff
@@ -149,9 +154,66 @@ executable geniusyield-market-maker-exe
     , vector
+  ghc-options:      -Wall
+  default-language: Haskell2010
+  mixins:
```
Is it needed?

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/geniusyield-market-maker.cabal:
```diff
@@ -149,9 +154,66 @@ executable geniusyield-market-maker-exe
+    , random-fu
+    , vector
+  hs-source-dirs:   .
   ghc-options:      -Wall
```
Likely not needed.

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/geniusyield-market-maker.cabal:
```diff
@@ -149,9 +154,66 @@ executable geniusyield-market-maker-exe
+                  , geniusyield-orderbot-framework:common
+                  , ghc-typelits-knownnat
+                  , maestro-sdk
+                  , plutus-ledger-api
```
Some of the imports seem unnecessary. Can check for warnings in `cabal build all --enable-tests` in new compilation.

ajuggler replied:
Done.  Eliminated:
    - MonadRandom
    - plutus-core
    - plutus-ledger-api
    - ply-core
    - random
    - random-fu
    - vector

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/MakerBot.hs:
```diff
@@ -115,19 +130,129 @@ buildAndSubmitActions user@User {uColl, uStakeCred} providers netId ua di = flip
+  runStateT
+    (evolveStrategy runStrategy mb netId providers pp di)
+    MBReady
+  return ()
```
Is `return ()` here needed?

ajuggler replied:
Done.

sourabhxyz replied:
In general, `void $ ...` is helpful in such scenarios.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/MakerBot.hs:
```diff
@@ -115,19 +130,129 @@ buildAndSubmitActions user@User {uColl, uStakeCred} providers netId ua di = flip
+          lift $ threadDelay respiteDelay
+
+    MBSpooked1 {..} → do
+      if mbs1Relax > pccAfterExitRelaxAim1 cfg then do
```
Maybe it should be `>=`, also need to check for other places where such inequality test is performed.

ajuggler replied:
Indeed; done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Strategies.hs:
```diff
@@ -199,213 +219,222 @@ fixedSpreadVsMarketPriceStrategy
+logPricesProviderFail providers pI = case pI of
+  PriceSourceFail es pps _ → do
+    gyLogWarning providers logNS $ "Some prices provider(s) failed: " ++ (show pps)
+    mapM_ (gyLogInfo providers logNS) es
```
Since we are in `gyLogWarning`, and it seems like a continuation of same message, let's also have it under warning severity.

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/MakerBot.hs:
```diff
@@ -115,19 +130,129 @@ buildAndSubmitActions user@User {uColl, uStakeCred} providers netId ua di = flip
+                     psf@PriceSourceFail {} → do
+                       lift $ logPricesProviderFail providers psf
+
+                     _                 → do
```
Let's have this constructor named here.

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/MakerBot.hs:
```diff
@@ -115,19 +130,129 @@ buildAndSubmitActions user@User {uColl, uStakeCred} providers netId ua di = flip
+                     PriceMismatch1    → do
+                       lift $ gyLogInfo providers logNS $ "Price mismatch persists"
+                       put MBSpooked1 { mbs1Relax = 0, mbs1Worse = mbs1Worse + 1 }
+
```
`threadDelay` appears to be missing here. Likewise for `PriceMismatch2`, `PriceUnavailable`. Please also check for at other suitable places.

ajuggler replied:
Done; indeed `threadDelay` was handled correctly here.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/MakerBot.hs:
```diff
@@ -115,19 +130,129 @@ buildAndSubmitActions user@User {uColl, uStakeCred} providers netId ua di = flip
+
+                     _                 → do
+                       lift $ gyLogInfo providers logNS $ "Apparent recovery of Prices Providers; waiting for observation period to elapse."
+                       put MBSpooked1 { mbs1Relax = mbs1Relax + 1, mbs1Worse = mbs1Worse }
```
Can omit `mbs1Worse = mbs1Worse`

ajuggler replied:
Done.

sourabhxyz replied:
Might be missing an extension, but fine.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/MakerBot.hs:
```diff
@@ -115,19 +130,129 @@ buildAndSubmitActions user@User {uColl, uStakeCred} providers netId ua di = flip
+                   lift $ threadDelay respiteDelay
+
+    MBSpooked2 {..} → do
+      if mbs2Relax > pccAfterExitRelaxAim2 cfg then do
```
>=?

ajuggler replied:
Indeed it should be `>=`.  Done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/MakerBot.hs:
```diff
@@ -115,19 +130,129 @@ buildAndSubmitActions user@User {uColl, uStakeCred} providers netId ua di = flip
+
+           case pe of
+             PriceMismatch1    → do
+               lift $ gyLogWarning providers logNS $ "Price mismatch persists"
```
Maybe add a namespace to clarify it's for `MBSpooked2`.

ajuggler replied:
Done.  Added context as part of the error string.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/MakerBot.hs:
```diff
@@ -115,19 +130,129 @@ buildAndSubmitActions user@User {uColl, uStakeCred} providers netId ua di = flip
+             psf@PriceSourceFail {} → do
+               lift $ logPricesProviderFail providers psf
+
+             _                 → do
```
Have this constructor named.

ajuggler replied:
Done.

sourabhxyz commented on sample-preprod-maker-bot-config-gens-v2.yaml:
Let's update the earlier sample configs.

ajuggler replied:
Done.

sourabhxyz commented on sample-preprod-maker-bot-config-gens-v2.yaml:
```diff
@@ -0,0 +1,108 @@
+  # Registered decimal places for the token. As an examples, for `ADA` it is 6 and for `SNEK` it is 0.
+  precision: 6
+# Boiler plate for Pre-production testnet.
+mbc_fp_nft_policy: compiled-scripts/minting-policy
```
This boiler plate was removed recently and so has to be omitted for.

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/test/mmbot.log:
I guess it's not needed, so better to be omitted.

ajuggler replied:
Done, this file and the "flags" have been deleted.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -52,11 +56,84 @@ import           Maestro.Types.V1
+  deriving (FromJSON) via CustomJSON '[FieldLabelModifier '[CamelToSnake]] PriceConfigV2
+
+data PriceCommonCfg = PriceCommonCfg
+  { pccCommonResolution       ∷ !CommonResolution,
```
Likely best to get rid of `CommonResolution` as it imposes the restriction for new price providers.

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -65,6 +142,285 @@ data MMToken = MMToken
+  
+  offset ← case mokcOffsetPath of
+    Nothing   → pure 0
+    Just path → readOffset path `catch` handleOffsetReadError
```
`handleOffsetReadError ` is throwing the caught exception and not doing anything else, in which case it is redundant.

ajuggler replied:
This was eliminated, as state is now managed with MVars.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices/Taptools.hs:
```diff
@@ -0,0 +1,95 @@
+
+-- | Taptools time resolution for OHLC Candles
+data TtResolution = TtRes3m | TtRes5m | TtRes15m | TtRes30m | TtRes1h | TtRes2h | TtRes4h | TtRes12h
+                  | TtRes1d | TtRes3d | TtRes1w  | TtRes1mo
```
Let's match with TapTools specification for it's provider: `TtRes1mo` -> `TtRes1M`.

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices/Taptools.hs:
```diff
@@ -0,0 +1,95 @@
+  resolutionInherited = Map.fromList [ (CRes5m, TtRes5m), (CRes15m, TtRes15m), (CRes30m, TtRes30m), (CRes1h, TtRes1h)
+                                     , (CRes4h, TtRes4h), (CRes1d, TtRes1d), (CRes1w, TtRes1w), (CRes1mo, TtRes1mo) ]
+
+type TtUnit = String
```
Have a look at how it was handled in PR at `dex-contracts-api`. Idea is to make it a newtype wrapper around `GYAssetClass`.

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices/Taptools.hs:
```diff
@@ -0,0 +1,95 @@
+  "token" :> "ohlcv" :> Header "x-api-key" Text
+                     :> QueryParam "unit" TtUnit
+                     :> QueryParam' '[Required, Strict] "interval" TtResolution
+                     :> QueryParam "numIntervals" Int
```
should be word64 or something to denote that negative values aren't possible.

ajuggler replied:
Done:  used `Word8` since we only need to accommodate for `numIntervals = 1`.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices/Taptools.hs:
```diff
@@ -0,0 +1,95 @@
+  , volume :: !Double
+  } deriving stock Show
+
+instance FromJSON TtOHLCV where
```
use automatic json derivation as described earlier.

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices/Taptools.hs:
```diff
@@ -0,0 +1,95 @@
+taptoolsManager = newManager tlsManagerSettings
+
+taptoolsBaseUrl :: BaseUrl
+taptoolsBaseUrl = BaseUrl Http "openapi.taptools.io" 80 "api/v1"
```
There is still `Http` here.

sourabhxyz replied:
Also for Https, port has to be different (443?).

sourabhxyz replied:
Again, this is being handled for in PR at dex-contracts-api side and can be referred there.

ajuggler replied:
Done:  it was as simple as changing the port to 443.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -52,11 +56,84 @@ import           Maestro.Types.V1
+
+data TaptoolsConfig = TaptoolsConfig
+  { ttcApiKey             ∷ !(Confidential Text),
+    ttcResolutionOverride ∷ !(Maybe TtResolution),
```
Once we get rid of common resolution, naming has to omit `Override` here.

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -65,6 +142,285 @@ data MMToken = MMToken
+
+data PricesProvidersAggregator = PPA
+  { ppaCommonCfg     ∷ !PriceCommonCfg,
+    ppaPricesCluster ∷ ![PricesProviderCfg]
```
Can be `NonEmpty`.

ajuggler replied:
Indeed; done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -65,6 +142,285 @@ data MMToken = MMToken
+          , pccRespiteDelayFactor     = 0.5     -- TODO: sensible value?
+          , pccAfterExitRelaxAim1     = 0       -- placeholder, unused
+          , pccAfterExitWorseMax1     = 0       -- placeholder, unused
+          , pccAfterExitRelaxAim2     = 10      -- TODO: sensible value?
```
Maybe `1`?

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -65,6 +142,285 @@ data MMToken = MMToken
+  deriving stock Show
+
+-- | Prices provider error:  `SourceUnavailable displayException pricesProvider`.
+data SourceError = SourceUnavailable String String
```
Can have tags for providers, like `data ProviderTag = MaestroTag | TapToolsTag`, so we can avoid `String` here. This is optional.

ajuggler replied:
Decided to keep it simple for now, and use String.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -65,6 +142,285 @@ data MMToken = MMToken
+            ('"':xs) | not (null xs) && last xs == '"' → init xs
+            _                                          → s
+
+  _         → throwIO $ userError "Only Mainnet is supported."
```
Preferably an exception type for TapTools.

sourabhxyz replied:
Perhaps could be part of earlier checks like `buildPP`?

ajuggler replied:
Done:  now check is part of `buildPP`.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices/Taptools.hs:
```diff
@@ -0,0 +1,95 @@
+taptoolsBaseUrl :: BaseUrl
+taptoolsBaseUrl = BaseUrl Http "openapi.taptools.io" 80 "api/v1"
+
+priceFromTaptools :: Confidential Text -> TtUnit -> TtResolution -> Int -> ClientEnv -> IO (Either ClientError [TtOHLCV])
```
API key is to be hidden.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -65,6 +142,285 @@ data MMToken = MMToken
+          Right [] → return . Left $ SourceUnavailable "Empty OHLCV." "Taptools"
+          Right (ttOHLCV : _) → do
+            let price         = close ttOHLCV
+                precisionDiff = 10 ** fromIntegral (mmtPrecision mmtLovelace - precision)
```
To be confirmed whether it is needed for TapTools with tokens like SNEK, AGIX.

ajuggler replied:
Done:  it turns out that asset's precision is irrelevant in the way price is provided by Taptools.  (So removed "adjustedPrice" logic for taptools.)

sourabhxyz replied:
It seems that this is required, please have a look.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -52,11 +56,84 @@ import           Maestro.Types.V1
+data MockConfig = MockConfig  -- For testing
+  { mokcPrice          ∷ !Double,
+    mokcOffsetPath     ∷ !(Maybe FilePath),
+    mokcAvailablePath  ∷ !(Maybe FilePath)
```
Replaced by a `MVar`, likewise for `mokcOffsetPath`.

ajuggler replied:
Done.  Now price in mock provider is handled by an `MVar (Maybe Double)`.  If `Just p` then `p` is the price; if `Nothing` then the provider is deemed "unavailable".

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -65,6 +142,285 @@ data MMToken = MMToken
+          priceAnalyzed
+            | rsd > ths2      = PriceMismatch2
+            | rsd > ths1      = PriceMismatch1
+            | hasSourceFailed = PriceSourceFail unavailableExceptions unavailablePProviders p
```
`unavailableExceptions` could be abstracted as a datatype listing all available providers exception, like in `GeniusYield.TxBuilder.Errors` module inside Atlas.

sourabhxyz commented on geniusyield-market-maker/test/Main.hs:
```diff
@@ -0,0 +1,73 @@
+  logRef  <- newMVar []
+
+  let action           = "Run"
+  let frameworkCfgPath = "../secrets/my-atlas-config-maestro.json"
```
Preferably make it an environment variable (which can be obtained using `getEnv`).

sourabhxyz replied:
Likewise for `mBotConfigFile`.

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/test/Main.hs:
```diff
@@ -0,0 +1,73 @@
+  mbc     <- readMBotConfig mBotConfigFile
+  mb      <- buildMakerBot mbc
+  di      <-
+    case cfgNetworkId coreCfg of
+      GYTestnetPreprod -> pure dexInfoDefaultPreprod
+      GYMainnet -> pure dexInfoDefaultMainnet
+      _ -> throwIO $ userError "Only Preprod and Mainnet are supported."
```
This can be abstracted for.

ajuggler replied:
Done; removed this code duplication in `test/Main.hs`.

sourabhxyz commented on geniusyield-market-maker/test/GeniusYield/Test/MarketMaker/Utils.hs:
```diff
@@ -0,0 +1,149 @@
+                -> GYLogNamespace -> GYLogSeverity -> String
+                -> IO ()
+augmentedLogRun logRef f ns s msg = do
+  modifyMVar_ logRef $ \current -> pure $ current ++ [LDLog (MBLog (s, ns, msg))]
```
Prepending is faster (O(1)) for Haskell prelude list.

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/test/Main.hs:
```diff
@@ -0,0 +1,73 @@
+      ++ show (addressToText $ addrUser (cfgNetworkId coreCfg) $ mbUser mb)
+
+    case action of
+      "Run"    -> do
```
`case action of` is seemingly not required, can simply call for `"Run"` action.

ajuggler replied:
Done.

sourabhxyz commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs:
```diff
@@ -52,11 +56,84 @@ import           Maestro.Types.V1
+  deriving (FromJSON) via CustomJSON '[FieldLabelModifier '[CamelToSnake]] TaptoolsConfig
+
+data MockConfig = MockConfig  -- For testing
+  { mokcPrice          ∷ !Double,
```
Instead of adding offset, this price itself could be a MVar.

ajuggler replied:
Indeed; done.

sourabhxyz commented on geniusyield-market-maker/test/GeniusYield/Test/MarketMaker/Utils.hs:
```diff
@@ -0,0 +1,149 @@
+  f ns s msg
+
+mbTest :: forall a. (PPStatus -> StateT a IO ()) -> StateT a IO ()
+mbTest f = foldM (const f) () testSequence
```
`mapM_`?

ajuggler replied:
Indeed; done.

ajuggler commented:
Experimentation shows that, with original definition of `taptoolsManager`,
```haskell
taptoolsManager :: Confidential Text -> IO Manager
taptoolsManager apiKey = newManager $ tlsManagerSettings { managerModifyRequest = withHeaders }
  where
    Confidential apiKey' = apiKey

    withHeaders :: Request -> IO Request
    withHeaders req = do
      return $ req { requestHeaders = ("x-api-key", encodeUtf8 apiKey') :
                                      filter (("x-api-key" /=) . fst) (requestHeaders req)
                   }
```
API key is automatically concealed.

ajuggler commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/MakerBot.hs:
```diff
@@ -115,19 +130,129 @@ buildAndSubmitActions user@User {uColl, uStakeCred} providers netId ua di = flip
+  let cfg          = ppaCommonCfg . pricesAggregatorPP $ pp
+      respiteDelay = ceiling $ pccRespiteDelayFactor cfg * fromIntegral mbDelay
+
+  case mbFret of
```
Done.

ajuggler commented on geniusyield-market-maker/src/GeniusYield/MarketMaker/Strategies.hs:
```diff
@@ -199,213 +219,222 @@ fixedSpreadVsMarketPriceStrategy
+flip4 f = \b c d a → f a b c d
+
+logPricesProviderFail ∷ GYProviders → PriceIndicator → IO ()
+logPricesProviderFail providers pI = case pI of
```
Done.

sourabhxyz commented on sample-preprod-maker-bot-config-gens.yaml:
```diff
@@ -3,35 +3,13 @@
+  # Load the wallet from mnemonic seed phrase.
   ur_mnemonic:
     mnemonic:
-      - health
```
Let's not remove this mnemonic.

sourabhxyz commented on sample-preprod-maker-bot-config-gens.yaml:
```diff
@@ -3,35 +3,13 @@
 
 # Wallet configuration.
 mbc_user:
-  # Either load wallet from signing key generated using `cardano-cli` (normal or extended).
```
Also, `ur_s_key_path` is not to be omitted.

sourabhxyz commented on geniusyield-market-maker/CHANGELOG.md:
```diff
@@ -1,5 +1,10 @@
 # Revision history for geniusyield-market-maker
 
+## 0.5.0 -- 2024-05-14
```
@ajuggler Need to mention this in cabal file as well.

sourabhxyz commented on README.md:
```diff
@@ -132,17 +132,34 @@ https://github.com/geniusyield/market-maker/blob/1c71faba6ff54c62cc8100eecbae69c
+* `mbc_price_config` gives the configuration on how to get market price using [Maestro](https://docs.gomaestro.org/DefiMarketAPI/mkt-dex-ohlc) or [Taptools](https://openapi.taptools.io/#tag/Market-Tokens/paths/~1token~1ohlcv/get) endpoints, for a token.
+  * `pc_price_common_cfg` contains the configuration parameters common to both prices providers.
+    * `pcc_network_id` determines Cardano network which is mentioned for in API calls. It should always be kept `mainnet` as of now.
+    * `pcc_price_diff_threshold1` if the *relative standard deviation*[^relstddev] among the prices providers is above this parameter, automatic cancelation of open orders is triggered (without raising logs' severity to "warning").
```
typo for `cancelation`.
